llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp

```diff
@@ -150,14 +150,11 @@ PropSetRegTy computeModuleProperties(const Module &M,
     PropSet.add(PropSetRegTy::SYCL_DEVICE_REQUIREMENTS,
                 computeDeviceRequirements(M, EntryPoints).asMap());
   }
-  auto *SpecConstsMD =
-      M.getNamedMetadata(SpecConstantsPass::SPEC_CONST_MD_STRING);
-  bool SpecConstsMet =
-      SpecConstsMD != nullptr && SpecConstsMD->getNumOperands() != 0;
-  if (SpecConstsMet) {
-    // extract spec constant maps per each module
-    SpecIDMapTy TmpSpecIDMap;
-    SpecConstantsPass::collectSpecConstantMetadata(M, TmpSpecIDMap);
+
+  // extract spec constant maps per each module
+  SpecIDMapTy TmpSpecIDMap;
+  SpecConstantsPass::collectSpecConstantMetadata(M, TmpSpecIDMap);
+  if (!TmpSpecIDMap.empty()) {
```

Comment:
I assume that both spec constants metadata and default values metadata are in sync, i.e. if one is empty then another is also empty. However, I wonder if we need to note that in a comment, or in form of some `assert` in `else` branch put under `#ifndef NDEBUG`?

I'm just looking for what others think, I do not have a strong opinion here, so not asking to do any changes just yet.

Reply:
That seems reasonable to me, let me add it.

Reply:
Running unit tests, it seems it actually is the case that `TmpSpecIDMap` is empty but `IsSpecConstantDefault` is true, and I'm not sure if there are any other cases that should bet rue.

Reply:
Merging for now, let me know if you have any other ideas on something we can assert, happy to implement it.

Reply:
> Running unit tests, it seems it actually is the case that `TmpSpecIDMap` is empty but `IsSpecConstantDefault` is true, and I'm not sure if there are any other cases that should bet rue.

What I meant is that `collectSpecConstantMetadata` and `collectSpecConstantMetadata` should either both return empty map, or should both return non-empty map. That's the implicit expectation we currently have.

I'm not sure what `IsSpecConstantDefault` is, but if that's a marker that device image contains default values of spec constants then it is also a weird state. That image should not be produced if there are no spec constants, because it will be no different from the original device image it was produced from. We even have an early exit in the corresponding function

Reply:
Sorry the two functions you mentioned are the same, did you mean `collectSpecConstantMetadata` and `collectSpecConstantDefaultValuesMetadata`?

Reply:
Right, copy-paste mistake

Reply:
that seems to work, pr coming soon. thanks

Reply:
https://github.com/intel/llvm/pull/15420